pkg/extensions/search/gql_proxy/gql_proxy.go

```diff
@@ -0,0 +1,105 @@
+package gqlproxy
+
+import (
+	"net/http"
+
+	"github.com/vektah/gqlparser/v2"
+	"github.com/vektah/gqlparser/v2/ast"
+
+	"zotregistry.dev/zot/pkg/api/config"
+	"zotregistry.dev/zot/pkg/api/constants"
+	"zotregistry.dev/zot/pkg/log"
+)
+
+type GqlScaleOutHandlerFunc func(*config.Config, http.ResponseWriter, *http.Request)
+
+// Returns a wrapped handler that can handle request proxying for GQL
+// queries when running in cluster mode without shared storage (each instance has its own metadata).
+// Requests are only proxied in local cluster mode as in this mode, each instance holds only the
+// metadata for the images that it serves, however, in shared storage mode,
+// all the instances have access to all the metadata so any can respond.
+func GqlProxyRequestHandler(
+	config *config.Config,
+	log log.Logger,
+	gqlSchema *ast.Schema,
+) func(handler http.Handler) http.Handler {
+	proxyFunctionalityMap := map[string]GqlScaleOutHandlerFunc{
```

Comment:
@rchincha this map statically stores which GQL operation needs which kind of handler.

There will be a couple of generic handlers such as fanout and some specific handlers if any of the operations need custom behavior.

What do you think about this approach?

It's better than last time since we don't need to maintain a separate handler for each operation type, but I'm open to more ideas on making this better.